### ISSUE TITLE
hubble-relay: failed to create gRPC client

### ISSUE THREAD

**Issue by alex1989hu** [cilium/cilium#16361]
## Bug report

After upgrading from Cilium `v1.9.7` to Cilium `v1.10.0` hubble-relay stopped working: it can not connect to `node:4244`.

I did not change anything in Helm Chart values.yaml file, except Chart version replace.

I thought issue caused :
- by `CiliumNetworkPolicy` -> removed -> nothing changed.
- relay tls server is needed, so enabled option `hubble.relay.tls.server.enabled`: true > rerun cronjob -> killed `hubble-relay` Pod -> nothing changed.

Sample log lines:
> level=info msg=Connecting address="3.249.66.182:4244" hubble-tls=true peer=prod-k8s-master-3 subsys=hubble-relay
level=warning msg="Failed to create gRPC client" address="3.249.66.182:4244" error="context deadline exceeded" hubble-tls=true next-try-in=10s peer=prod-k8s-master-3 subsys=hubble-relay

`next-try-in` is constantly increasing but nothing helped.

**General Information**

- Cilium version (run `cilium version`)
> Client: 1.10.0 952d9d3 2021-05-19T18:42:32+02:00 go version go1.16.4 linux/amd64
Daemon: 1.10.0 952d9d3 2021-05-19T18:42:32+02:00 go version go1.16.4 linux/amd64

- Kernel version (run `uname -a`)
> Linux prod-k8s-master-3 5.10.29-talos #1 SMP Wed Apr 28 14:43:26 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
- Orchestration system version in use (e.g. `kubectl version`, ...)
> Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.1", GitCommit:"5e58841cce77d4bc13713ad2b91fa0d961e69192", GitTreeState:"clean", BuildDate:"2021-05-12T14:18:45Z", GoVersion:"go1.16.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.1", GitCommit:"5e58841cce77d4bc13713ad2b91fa0d961e69192", GitTreeState:"clean", BuildDate:"2021-05-12T14:12:29Z", GoVersion:"go1.16.4", Compiler:"gc", Platform:"linux/amd64"}

- Link to relevant artifacts (policies, deployments scripts, ...)
N/A

- Generate and upload a system zip:
```
curl -sLO https://git.io/cilium-sysdump-latest.zip && python cilium-sysdump-latest.zip
```

Can not attach cilium-sysdump - file size is `52M`

**How to reproduce the issue**

Do not know.

**Comment by rolinh** [comment]
I am unable to reproduce. Could you please share your Helm install options (at least the Hubble section) as well as the upgrade command?

**Comment by alex1989hu** [comment]
Sure, @rolinh 

Cilium installation and upgrading is managed by [ArgoCD](https://argo-cd.readthedocs.io/en/stable/).

Chart.yaml

```
apiVersion: v2
appVersion: "1.0"
description: Helm chart for Cilium
name: cilium
version: 0.0.1
dependencies:
- name: cilium
  version: 1.10.0
  repository: "@cilium"
```
values.yaml

```yaml
cilium:
  bpf:
    policyMapMax: 40960
  hubble:
    enabled: true
    metrics:
      enabled:
      - drop:sourceContext=pod;destinationContext=pod
      - icmp
      - tcp
      - flow
    relay:
      enabled: true
    serviceMonitor:
      enabled: true
    tls:
      enabled: true
      auto:
        enabled: true
        method: cronJob
        certValidityDuration: 70
        schedule: "0 0 1 */2 *"
    ui:
      enabled: true
      backend:
        resources:
          limits:
            cpu: 1000m
            memory: 128Mi
          requests:
            cpu: 100m
            memory: 32Mi
      frontend:
        resources:
          limits:
            cpu: 1000m
            memory: 160Mi
          requests:
            cpu: 100m
            memory: 96Mi
  operator:
    replicas: 2
    prometheus:
      enabled: true
      serviceMonitor:
        enabled: true
  prometheus:
    enabled: true
    serviceMonitor:
      enabled: true
```

**Comment by alex1989hu** [comment]
[hubble-relay-557d7cb747-tm9br-hubble-relay-20210528-185956.log](https://github.com/cilium/cilium/files/6562704/hubble-relay-557d7cb747-tm9br-hubble-relay-20210528-185956.log)

**Comment by rolinh** [comment]
Ah, I see that you're using the cronjob method for cert generation, I'll try to reproduce again.
Did you encounter any issue with Helm during the upgrade? Asking because issue #16316 might be related.

**Comment by alex1989hu** [comment]
ArgoCD did not report any issue during Cilium upgrade.

Here is the current manifest of cronjob method for cert generation:

```yaml
apiVersion: batch/v1
kind: CronJob
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"batch/v1beta1","kind":"CronJob","metadata":{"annotations":{},"labels":{"app.kubernetes.io/instance":"cilium","k8s-app":"hubble-generate-certs"},"name":"hubble-generate-certs","namespace":"kube-system"},"spec":{"concurrencyPolicy":"Forbid","jobTemplate":{"spec":{"template":{"metadata":{"labels":{"k8s-app":"hubble-generate-certs"}},"spec":{"containers":[{"args":["--cilium-namespace=kube-system","--hubble-ca-reuse-secret=true","--hubble-ca-secret-name=hubble-ca-secret","--hubble-ca-generate=true","--hubble-ca-validity-duration=6048000s","--hubble-ca-config-map-create=true","--hubble-ca-config-map-name=hubble-ca-cert","--hubble-server-cert-generate=true","--hubble-server-cert-common-name=*.default.hubble-grpc.cilium.io","--hubble-server-cert-validity-duration=6048000s","--hubble-server-cert-secret-name=hubble-server-certs","--hubble-relay-client-cert-generate=true","--hubble-relay-client-cert-validity-duration=6048000s","--hubble-relay-client-cert-secret-name=hubble-relay-client-certs","--hubble-relay-server-cert-generate=true","--hubble-relay-server-cert-validity-duration=6048000s","--hubble-relay-server-cert-secret-name=hubble-relay-server-certs"],"command":["/usr/bin/cilium-certgen"],"image":"quay.io/cilium/certgen:v0.1.4","imagePullPolicy":"IfNotPresent","name":"certgen"}],"hostNetwork":true,"restartPolicy":"OnFailure","serviceAccount":"hubble-generate-certs","serviceAccountName":"hubble-generate-certs"}},"ttlSecondsAfterFinished":1800}},"schedule":"0 0 1 */2 *"}}
  creationTimestamp: "2021-05-28T16:38:55Z"
  labels:
    app.kubernetes.io/instance: cilium
    k8s-app: hubble-generate-certs
  name: hubble-generate-certs
  namespace: kube-system
  resourceVersion: "97540106"
  uid: bb38222a-26b4-4836-a082-97cc1fe9153b
spec:
  concurrencyPolicy: Forbid
  failedJobsHistoryLimit: 1
  jobTemplate:
    metadata:
      creationTimestamp: null
    spec:
      template:
        metadata:
          creationTimestamp: null
          labels:
            k8s-app: hubble-generate-certs
        spec:
          containers:
          - args:
            - --cilium-namespace=kube-system
            - --hubble-ca-reuse-secret=true
            - --hubble-ca-secret-name=hubble-ca-secret
            - --hubble-ca-generate=true
            - --hubble-ca-validity-duration=6048000s
            - --hubble-ca-config-map-create=true
            - --hubble-ca-config-map-name=hubble-ca-cert
            - --hubble-server-cert-generate=true
            - --hubble-server-cert-common-name=*.default.hubble-grpc.cilium.io
            - --hubble-server-cert-validity-duration=6048000s
            - --hubble-server-cert-secret-name=hubble-server-certs
            - --hubble-relay-client-cert-generate=true
            - --hubble-relay-client-cert-validity-duration=6048000s
            - --hubble-relay-client-cert-secret-name=hubble-relay-client-certs
            - --hubble-relay-server-cert-generate=true
            - --hubble-relay-server-cert-validity-duration=6048000s
            - --hubble-relay-server-cert-secret-name=hubble-relay-server-certs
            command:
            - /usr/bin/cilium-certgen
            image: quay.io/cilium/certgen:v0.1.4
            imagePullPolicy: IfNotPresent
            name: certgen
            resources: {}
            terminationMessagePath: /dev/termination-log
            terminationMessagePolicy: File
          dnsPolicy: ClusterFirst
          hostNetwork: true
          restartPolicy: OnFailure
          schedulerName: default-scheduler
          securityContext: {}
          serviceAccount: hubble-generate-certs
          serviceAccountName: hubble-generate-certs
          terminationGracePeriodSeconds: 30
      ttlSecondsAfterFinished: 1800
  schedule: 0 0 1 */2 *
  successfulJobsHistoryLimit: 3
  suspend: false
status: {}
```

I can even start it manually:

```bash
kubectl -n kube-system create job manual-cert --from=cronjob/hubble-generate-certs
job.batch/manual-cert created
kubectl -n kube-system logs manual-cert-g47rh
time="2021-05-31T08:41:46Z" level=info msg="cilium-certgen 0.1.4" subsys=cilium-certgen
time="2021-05-31T08:41:46Z" level=info msg="Loaded Hubble CA Secret" subsys=cilium-certgen
time="2021-05-31T08:41:46Z" level=info msg="Generating server certificates for Hubble" subsys=cilium-certgen
time="2021-05-31T08:41:46Z" level=info msg="Creating CSR for certificate" certCommonName="*.default.hubble-grpc.cilium.io" certUsage="[signing key encipherment server auth]" certValidityDuration=1680h0m0s subsys=generate
time="2021-05-31T08:41:46Z" level=info msg="generate received request" subsys=cfssl syslog=info
time="2021-05-31T08:41:46Z" level=info msg="received CSR" subsys=cfssl syslog=info
time="2021-05-31T08:41:46Z" level=info msg="generating key: ecdsa-256" subsys=cfssl syslog=info
time="2021-05-31T08:41:46Z" level=info msg="encoded CSR" subsys=cfssl syslog=info
time="2021-05-31T08:41:46Z" level=info msg="signed certificate with serial number 635563111769063360358372746673330085496126630961" subsys=cfssl syslog=info
time="2021-05-31T08:41:46Z" level=info msg="Generating client certificates for Hubble Relay" subsys=cilium-certgen
time="2021-05-31T08:41:46Z" level=info msg="Creating CSR for certificate" certCommonName="*.hubble-relay.cilium.io" certUsage="[signing key encipherment server auth client auth]" certValidityDuration=1680h0m0s subsys=generate
time="2021-05-31T08:41:46Z" level=info msg="generate received request" subsys=cfssl syslog=info
time="2021-05-31T08:41:46Z" level=info msg="received CSR" subsys=cfssl syslog=info
time="2021-05-31T08:41:46Z" level=info msg="generating key: ecdsa-256" subsys=cfssl syslog=info
time="2021-05-31T08:41:46Z" level=info msg="encoded CSR" subsys=cfssl syslog=info
time="2021-05-31T08:41:46Z" level=info msg="signed certificate with serial number 675871173507915508298520737236522438620418419498" subsys=cfssl syslog=info
time="2021-05-31T08:41:46Z" level=info msg="Generating server certificates for Hubble Relay" subsys=cilium-certgen
time="2021-05-31T08:41:46Z" level=info msg="Creating CSR for certificate" certCommonName="*.hubble-relay.cilium.io" certUsage="[signing key encipherment server auth]" certValidityDuration=1680h0m0s subsys=generate
time="2021-05-31T08:41:46Z" level=info msg="generate received request" subsys=cfssl syslog=info
time="2021-05-31T08:41:46Z" level=info msg="received CSR" subsys=cfssl syslog=info
time="2021-05-31T08:41:46Z" level=info msg="generating key: ecdsa-256" subsys=cfssl syslog=info
time="2021-05-31T08:41:46Z" level=info msg="encoded CSR" subsys=cfssl syslog=info
time="2021-05-31T08:41:46Z" level=info msg="signed certificate with serial number 467479436155931075501463104458300260112837948668" subsys=cfssl syslog=info
time="2021-05-31T08:41:46Z" level=info msg="Creating K8s ConfigMap" k8sConfigMapName=hubble-ca-cert k8sConfigMapNamespace=kube-system subsys=generate
time="2021-05-31T08:41:46Z" level=info msg="ConfigMap already exists, updating it instead" k8sConfigMapName=hubble-ca-cert k8sConfigMapNamespace=kube-system subsys=generate
time="2021-05-31T08:41:46Z" level=info msg="Creating K8s Secret" k8sSecretName=hubble-server-certs k8sSecretNamespace=kube-system subsys=generate
time="2021-05-31T08:41:46Z" level=info msg="Secret already exists, updating it instead" k8sSecretName=hubble-server-certs k8sSecretNamespace=kube-system subsys=generate
time="2021-05-31T08:41:46Z" level=info msg="Creating K8s Secret" k8sSecretName=hubble-relay-client-certs k8sSecretNamespace=kube-system subsys=generate
time="2021-05-31T08:41:46Z" level=info msg="Secret already exists, updating it instead" k8sSecretName=hubble-relay-client-certs k8sSecretNamespace=kube-system subsys=generate
time="2021-05-31T08:41:46Z" level=info msg="Creating K8s Secret" k8sSecretName=hubble-relay-server-certs k8sSecretNamespace=kube-system subsys=generate
time="2021-05-31T08:41:46Z" level=info msg="Secret already exists, updating it instead" k8sSecretName=hubble-relay-server-certs k8sSecretNamespace=kube-system subsys=generate
time="2021-05-31T08:41:46Z" level=info msg="Successfully generated all 4 requested certificates." subsys=cilium-certgen
```

**Comment by rolinh** [comment]
Interestingly, I fail to reproduce even when using the `values.yaml` file that you provided. If you `kubectl delete` the cronjob, does it fix the issue?

**Comment by rolinh** [comment]
@alex1989hu Could you please also provide a sysdump of the system, maybe host it elsewhere if it's too large for github?

**Comment by alex1989hu** [comment]
@rolinh: I can not attach here.

> Can not attach cilium-sysdump - file size is 52M

I have just uploaded (temporary) it here https://drive.google.com/file/d/1rN4Qc8Yi2BaT3ooBqqwMh_XFOGCWmOaO/view?usp=sharing

**Comment by rolinh** [comment]
@alex1989hu I realized that we do not include the Hubble Relay configmap in the sysdump. Could you please provide it as well?

**Comment by alex1989hu** [comment]
@rolinh sure

```yaml
# kubectl -n kube-system get configmaps hubble-relay-config -o yaml
apiVersion: v1
data:
  config.yaml: "peer-service: unix:///var/run/cilium/hubble.sock\nlisten-address:
    :4245\ndial-timeout: \nretry-timeout: \nsort-buffer-len-max: \nsort-buffer-drain-timeout:
    \ntls-client-cert-file: /var/lib/hubble-relay/tls/client.crt\ntls-client-key-file:
    /var/lib/hubble-relay/tls/client.key\ntls-hubble-server-ca-files: /var/lib/hubble-relay/tls/hubble-server-ca.crt\ntls-server-cert-file:
    /var/lib/hubble-relay/tls/server.crt\ntls-server-key-file: /var/lib/hubble-relay/tls/server.key\n"
kind: ConfigMap
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","data":{"config.yaml":"peer-service: unix:///var/run/cilium/hubble.sock\nlisten-address: :4245\ndial-timeout: \nretry-timeout: \nsort-buffer-len-max: \nsort-buffer-drain-timeout: \ntls-client-cert-file: /var/lib/hubble-relay/tls/client.crt\ntls-client-key-file: /var/lib/hubble-relay/tls/client.key\ntls-hubble-server-ca-files: /var/lib/hubble-relay/tls/hubble-server-ca.crt\ntls-server-cert-file: /var/lib/hubble-relay/tls/server.crt\ntls-server-key-file: /var/lib/hubble-relay/tls/server.key\n"},"kind":"ConfigMap","metadata":{"annotations":{},"labels":{"app.kubernetes.io/instance":"cilium"},"name":"hubble-relay-config","namespace":"kube-system"}}
  creationTimestamp: "2021-04-22T08:06:30Z"
  labels:
    app.kubernetes.io/instance: cilium
  name: hubble-relay-config
  namespace: kube-system
  resourceVersion: "97533158"
  uid: 75235e9a-7877-4980-b1b0-b916150b0603
```

**Comment by rolinh** [comment]
> ArgoCD did not report any issue during Cilium upgrade.

This is really curious as I'm unable to upgrade from v1.9 to v1.10 when `hubble.tls.auto.method=cronJob` due to #16316, no matter if I use `helm upgrade` or `helm template + kubectl apply`.

> relay tls server is needed, so enabled option hubble.relay.tls.server.enabled: true > rerun cronjob -> killed hubble-relay Pod -> nothing changed.

To be clear, Relay TLS server is not needed unless you want to enable mTLS for Relay itself. I see that in your cronjob spec, it is set to generate certs for Relay server (this is not required but it shouldn't affect the problem you're seeing):
```
            - --hubble-relay-server-cert-generate=true
            - --hubble-relay-server-cert-validity-duration=6048000s
            - --hubble-relay-server-cert-secret-name=hubble-relay-server-certs
```

Have you tried to re-run the cronjob with a command like this one? It is perfectly OK to re-generate the TLS certs, they will be picked up automatically by Hubble server and Relay as these components support TLS hot-reloading.
```
kubectl -n kube-system create job --from=cronjob/hubble-generate-certs hubble-generate-certs-manual-001 
```

**Comment by alex1989hu** [comment]
> Have you tried to re-run the cronjob with a command like this one?

Yes, see the last code snippet:
https://github.com/cilium/cilium/issues/16361#issuecomment-851326945

**Comment by rolinh** [comment]
> > Have you tried to re-run the cronjob with a command like this one?
> 
> Yes, see the last code snippet:

I missed this somehow :facepalm: 

So the job run looks OK to me. I also see that the cilium agent seem to pick the update:
```
2021-05-28T16:39:49.727070277Z level=info msg="Certificate authority updated" config=tls-server subsys=hubble
2021-05-28T16:39:49.727253174Z level=info msg="Keypair updated" config=tls-server keypair-sn=40ec90080f4e117d32d99d71221e84ca0c9c1304 subsys=hubble
```

I'd expect to see a similar message in the Hubble Relay logs but I don't. I guess, based on the logs timestamp, that you the Relay instance was re-created after the certs generation so it should have picked the new certificates as well.

Could you please run `hubble status` (`hubble list nodes -o wide` could be useful as well) against Relay and post the output?

**Comment by alex1989hu** [comment]
I can execute commands in `cilium-agent` container, `hubble-relay` container does not have shell:

![image](https://user-images.githubusercontent.com/10764100/120355555-ad047200-c303-11eb-95c4-cc528c7ca523.png)

**Comment by rolinh** [comment]
@alex1989hu I meant to target the Hubble Relay service using the CLI. This does not require exec'ing into a pod. Please, refer to [the documentation](https://docs.cilium.io/en/v1.10/operations/troubleshooting/#observing-flows-with-hubble-relay) for how to do this.

**Comment by alex1989hu** [comment]
@rolinh sorry. Based on the documentation I downloaded `hubble` CLI and set up a port-forward. No luck.

Actually, this is `v1.10`:

![image](https://user-images.githubusercontent.com/10764100/120361012-a8db5300-c309-11eb-8dd7-42f227a4e8e1.png)


```bash
$ kubectl -n kube-system port-forward service/hubble-relay --address 0.0.0.0 --address :: 4245:80
error: Service hubble-relay does not have a service port 80

 $ kubectl -n kube-system get svc hubble-relay
NAME           TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)   AGE
hubble-relay   ClusterIP   10.107.72.50   <none>        443/TCP   40d

$ kubectl -n kube-system port-forward service/hubble-relay --address 0.0.0.0 --address :: 4245:443
Forwarding from 0.0.0.0:4245 -> 4245
Forwarding from [::]:4245 -> 4245

# Open new terminal
$ nc -Czv localhost 4245
Connection to localhost 4245 port [tcp/*] succeeded!

$ hubble list nodes
failed to connect to 'localhost:4245': context deadline exceeded: connection closed
```

**Comment by rolinh** [comment]
@alex1989hu This fail because you enabled TLS for the Hubble Relay service itself (which is why it needs to be accessed on port 443 and not 80 as the default). You would need to setup your CLI with appropriate certs and CA to connect to Relay. However, to troubleshoot this issue it would be easier if this is disabled, i.e. `hubble.relay.tls.server.enabled=false`.

**Comment by alex1989hu** [comment]
> @alex1989hu This fail because you enabled TLS for the Hubble Relay service itself (which is why it needs to be accessed on port 443 and not 80 as the default). You would need to setup your CLI with appropriate certs and CA to connect to Relay. However, to troubleshoot this issue it would be easier if this is disabled, i.e. `hubble.relay.tls.server.enabled=false`.

Gotcha, I put `tls-allow-insecure: true` into `$HOME/.config/hubble/config.yaml` solves the issue, thus I do not have to disable TLS for the Hubble Relay service.

It does not look good.

```bash
 $ hubble status
Healthcheck (via localhost:4245): Ok
Current/Max Flows: 0/0
Flows/s: N/A
Connected Nodes: 0/9
Unavailable Nodes: 9
  - prod-k8s-master-1
  - prod-k8s-master-2
  - prod-k8s-master-3
  - prod-k8s-worker-8da7da49
  - prod-k8s-worker-8ddbacc3
  - prod-k8s-worker-f08f772d
  - prod-k8s-worker-fc988125
  - prod-k8s-worker-fdd18af8
  - prod-k8s-worker-gmjqaj86

$ hubble list nodes -o wide
NAME                       STATUS        AGE   FLOWS/S   CURRENT/MAX-FLOWS   VERSION   ADDRESS             TLS
prod-k8s-master-1          Unavailable   N/A   N/A       N/A                 N/A       3.249.66.180:4244   Enabled
prod-k8s-master-2          Unavailable   N/A   N/A       N/A                 N/A       3.249.66.181:4244   Enabled
prod-k8s-master-3          Unavailable   N/A   N/A       N/A                 N/A       3.249.66.182:4244   Enabled
prod-k8s-worker-8da7da49   Unavailable   N/A   N/A       N/A                 N/A       3.249.66.186:4244   Enabled
prod-k8s-worker-8ddbacc3   Unavailable   N/A   N/A       N/A                 N/A       3.249.66.188:4244   Enabled
prod-k8s-worker-f08f772d   Unavailable   N/A   N/A       N/A                 N/A       3.249.66.187:4244   Enabled
prod-k8s-worker-fc988125   Unavailable   N/A   N/A       N/A                 N/A       3.249.66.185:4244   Enabled
prod-k8s-worker-fdd18af8   Unavailable   N/A   N/A       N/A                 N/A       3.249.66.184:4244   Enabled
prod-k8s-worker-gmjqaj86   Unavailable   N/A   N/A       N/A                 N/A       3.249.66.189:4244   Enabled
```

**Comment by rolinh** [comment]
Troubleshooting why Hubble Relay fails to connect to peers is hard because of a limitation of the gRPC version that ships with Cilium. Would you mind trying out to deploy a Hubble CLI instance using the file in PR #16459? Then, run this command and paste the error you obtain:
```
kubectl exec -n kube-system deployment/hubble-cli -- \
    hubble observe --server 'tls://3.249.66.186:4244' \
        --tls-server-name 'prod-k8s-worker-8da7da49.default.hubble-grpc.cilium.io' \
        --tls-ca-cert-files /var/lib/hubble-relay/tls/hubble-server-ca.crt \
        --tls-client-cert-file /var/lib/hubble-relay/tls/client.crt \
        --tls-client-key-file /var/lib/hubble-relay/tls/client.key
```

**Comment by alex1989hu** [comment]
I have just deployed `hubble-cli` from the referenced PR.

`hubble-server-ca` expired, see later.

```bash
kubectl -n kube-system get deployment hubble-cli
NAME         READY   UP-TO-DATE   AVAILABLE   AGE
hubble-cli   1/1     1            1           2m40s
```

Executed the command:

```bash
failed to connect to 'tls://3.249.66.186:4244': context deadline exceeded: connection error: desc = "transport: authentication handshake failed: x509: certificate has expired or is not yet valid: current time 2021-06-10T17:04:20Z is after 2021-05-17T10:57:00Z"
command terminated with exit code 1
```

Its weird, it looks like the certificate file differs what node on the given port serves:

```bash
openssl s_client -connect 3.249.66.186:4244 -prexit -showcerts </dev/null 2>/dev/null

Certificate chain
 0 s:CN = *.default.hubble-grpc.cilium.io
   i:CN = hubble-ca.cilium.io
-----BEGIN CERTIFICATE-----
MIIB7TCCAZOgAwIBAgIUfGrvF2jgL4qFo6To4uO7rIgbTUQwCgYIKoZIzj0EAwIw
HjEcMBoGA1UEAxMTaHViYmxlLWNhLmNpbGl1bS5pbzAeFw0yMTA2MTAxNjQ0MDBa
Fw0yMTA4MTkxNjQ0MDBaMCoxKDAmBgNVBAMMHyouZGVmYXVsdC5odWJibGUtZ3Jw
Yy5jaWxpdW0uaW8wWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAAQUVz9Ilnl3JD36
X/QVxIDWAJOFmMNzo4Hefdbrt0QeyvBk1CwVuDLvYmmsmt8//WeDU2VcACAzK0zC
DagJ7Gw2o4GiMIGfMA4GA1UdDwEB/wQEAwIFoDATBgNVHSUEDDAKBggrBgEFBQcD
ATAMBgNVHRMBAf8EAjAAMB0GA1UdDgQWBBSmMyJyXF9yyyCMTiVjl75ms7jg9TAf
BgNVHSMEGDAWgBRARRrWcakVxc/g832SDBkCpmge/TAqBgNVHREEIzAhgh8qLmRl
ZmF1bHQuaHViYmxlLWdycGMuY2lsaXVtLmlvMAoGCCqGSM49BAMCA0gAMEUCIEiQ
HcC8bqpX3bDWhKL8XAPO+rKvHi7LbQPOdDcggO2oAiEAphl05pHlrdZYa8UqgLpq
BSvROcHt1mUzB8fwqXYiaGQ=
-----END CERTIFICATE-----
```

Let's see the certificate one by one.

**Expired** CA certificate is: `/var/lib/hubble-relay/tls/hubble-server-ca.crt`

```bash
-----BEGIN CERTIFICATE-----
MIIBgTCCASagAwIBAgIUEQcrOEugoWwyfIaRDcO5O8rzB/0wCgYIKoZIzj0EAwIw
HjEcMBoGA1UEAxMTaHViYmxlLWNhLmNpbGl1bS5pbzAeFw0yMTAzMDgxMDU3MDBa
Fw0yMTA1MTcxMDU3MDBaMB4xHDAaBgNVBAMTE2h1YmJsZS1jYS5jaWxpdW0uaW8w
WTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAASeGWKjcjcQ52agSjZF2sb3WGrA2E9S
0UCnEDWsY4JadCkbMcUOt1v5G6v50Zbe1+8pFNo9YgNhC6qJ1kobBHL2o0IwQDAO
BgNVHQ8BAf8EBAMCAQYwDwYDVR0TAQH/BAUwAwEB/zAdBgNVHQ4EFgQUQEUa1nGp
FcXP4PN9kgwZAqZoHv0wCgYIKoZIzj0EAwIDSQAwRgIhAPp3fDkT6aZwXeEBd96H
l8GlstjrTeNW5rwZiTZ2CBmqAiEArjPFupdZ8RW4IbECFMEXNGpve/72f4U8lcvB
eidxPtU=
-----END CERTIFICATE-----
```

Probably it causes the issue. I don't understand why it is expired.

To give you full picture, here is the client certificate which is fine: `/var/lib/hubble-relay/tls/client.crt`

```bash
-----BEGIN CERTIFICATE-----
MIIB6TCCAY+gAwIBAgIUZsPHkmPWMNla4khfjQbuXTMpdDUwCgYIKoZIzj0EAwIw
HjEcMBoGA1UEAxMTaHViYmxlLWNhLmNpbGl1bS5pbzAeFw0yMTA2MTAxNjQ0MDBa
Fw0yMTA4MTkxNjQ0MDBaMCMxITAfBgNVBAMMGCouaHViYmxlLXJlbGF5LmNpbGl1
bS5pbzBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IABCuB3DCSn/2DyI9LgjCU7BoC
7pPn7uHZywdH5lOVbiF+YLXGAwCCOgPVF5cUlk848+uajOisRxU8OcMGyp433sqj
gaUwgaIwDgYDVR0PAQH/BAQDAgWgMB0GA1UdJQQWMBQGCCsGAQUFBwMBBggrBgEF
BQcDAjAMBgNVHRMBAf8EAjAAMB0GA1UdDgQWBBROQNy/f56su9aiutyhOT6L8AbD
pzAfBgNVHSMEGDAWgBRARRrWcakVxc/g832SDBkCpmge/TAjBgNVHREEHDAaghgq
Lmh1YmJsZS1yZWxheS5jaWxpdW0uaW8wCgYIKoZIzj0EAwIDSAAwRQIhAPGbl1l3
VzK4TB8LM4h7YuBg7yUN0kqy+zZVyujd7xGUAiBrdbsiJOMtk1qkWclimLIXNt61
tQeuKPbfVemVxi1wbQ==
-----END CERTIFICATE-----
```

**Comment by rolinh** [comment]
@alex1989hu I think I know what's going on. Could you try editing the `hubble-generate-certs` cronjob to remove the following line from `Args`?
 ```
 - --hubble-ca-reuse-secret=true
 ```
 Then trigger the job again and see if it fixes the issue (I think it should generate new CA certs).

**Comment by alex1989hu** [comment]
@rolinh, I copied the original `CronJob`, removed that arg. Unfortunately, there is permission issue:

> time="2021-06-11T13:20:33Z" level=fatal msg="failed to generate certificates" error="failed to create secret for Hubble CA: secrets \"hubble-ca-secret\" is forbidden: User \"system:serviceaccount:kube-system:hubble-generate-certs\" cannot update resource \"secrets\" in API group \"\" in the namespace \"kube-system\"" subsys=cilium-certgen

Meaning that the Helm chart has a defect:

https://github.com/cilium/cilium/blob/952d9d33740fc77e1a94d5a52b38d2a313e5c570/install/kubernetes/cilium/templates/hubble-generate-certs-clusterrole.yaml#L32-L39

Here is the Pod log:

```bash
time="2021-06-11T13:20:33Z" level=info msg="cilium-certgen 0.1.4" subsys=cilium-certgen
time="2021-06-11T13:20:33Z" level=info msg="Creating CSR for certificate authority" certCommonName=hubble-ca.cilium.io certValidityDuration=1680h0m0s subsys=generate
time="2021-06-11T13:20:33Z" level=info msg="generate received request" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="received CSR" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="generating key: ecdsa-256" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="encoded CSR" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="signed certificate with serial number 124959461781947435713081743050694512010640070268" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="Generating server certificates for Hubble" subsys=cilium-certgen
time="2021-06-11T13:20:33Z" level=info msg="Creating CSR for certificate" certCommonName="*.default.hubble-grpc.cilium.io" certUsage="[signing key encipherment server auth]" certValidityDuration=1680h0m0s subsys=generate
time="2021-06-11T13:20:33Z" level=info msg="generate received request" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="received CSR" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="generating key: ecdsa-256" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="encoded CSR" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="signed certificate with serial number 22198575201069669743518100143765142861508975469" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="Generating client certificates for Hubble Relay" subsys=cilium-certgen
time="2021-06-11T13:20:33Z" level=info msg="Creating CSR for certificate" certCommonName="*.hubble-relay.cilium.io" certUsage="[signing key encipherment server auth client auth]" certValidityDuration=1680h0m0s subsys=generate
time="2021-06-11T13:20:33Z" level=info msg="generate received request" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="received CSR" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="generating key: ecdsa-256" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="encoded CSR" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="signed certificate with serial number 608458545455476931839087825516408982269593816752" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="Generating server certificates for Hubble Relay" subsys=cilium-certgen
time="2021-06-11T13:20:33Z" level=info msg="Creating CSR for certificate" certCommonName="*.hubble-relay.cilium.io" certUsage="[signing key encipherment server auth]" certValidityDuration=1680h0m0s subsys=generate
time="2021-06-11T13:20:33Z" level=info msg="generate received request" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="received CSR" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="generating key: ecdsa-256" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="encoded CSR" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="signed certificate with serial number 609364381472842499488164491968124808240241464862" subsys=cfssl syslog=info
time="2021-06-11T13:20:33Z" level=info msg="Creating K8s Secret" k8sSecretName=hubble-ca-secret k8sSecretNamespace=kube-system subsys=generate
time="2021-06-11T13:20:33Z" level=info msg="Secret already exists, updating it instead" k8sSecretName=hubble-ca-secret k8sSecretNamespace=kube-system subsys=generate
time="2021-06-11T13:20:33Z" level=fatal msg="failed to generate certificates" error="failed to create secret for Hubble CA: secrets \"hubble-ca-secret\" is forbidden: User \"system:serviceaccount:kube-system:hubble-generate-certs\" cannot update resource \"secrets\" in API group \"\" in the namespace \"kube-system\"" subsys=cilium-certgen
```

**Comment by joestringer** [comment]
@alex1989hu now that #16508 is fixed in Cilium 1.10.2 or later, are you still observing an issue?

**Comment by aanm** [comment]
This issue has not seen any activity since it was marked stale. Closing. Feel free to re-open a new issue if it is still persisting. Thank you.
